src/main/java/bio/terra/common/migirate/LiquibaseMigrator.java

```diff
@@ -0,0 +1,84 @@
+import liquibase.resource.ClassLoaderResourceAccessor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+/**
+ * The MigrateService invokes Liquibase to configure the service database. It provides the methods
+ * to clear stale locks, drop all data, and update the service database. We implement this in code,
+ * as opposed to using hooks in the Spring framework. We expect MC Terra services to scale
+ * horizontally and to perform rolling upgrades of the running instances. Therefore, we do not want
+ * the act of running an instance to suddenly upgrade the database schema.
+ */
+@Component
+public class LiquibaseMigrator {
+  private Logger logger = LoggerFactory.getLogger(LiquibaseMigrator.class);
+
+  /**
+   * Initialize drops existing tables in the database and reinitializes it with the changeset. This
+   * is useful when developing or running integration tests, where there is no expectation that the
+   * current state of the database is related to our changeset state.
+   *
+   * @param changesetFile - relative path to the changeset file in the project
+   * @param dataSource - database to operate on
+   */
+  public void initialize(String changesetFile, DataSource dataSource) {
+    migrateDatabase(changesetFile, dataSource, true);
+  }
+
+  /**
+   * Upgrade applies changesets to move the current state of the database to a new state. This is
+   * useful for alpha, staging, and production environments where we make some guarantee that the
+   * database schema is in a valid state for applying changesets to.
+   *
+   * @param changesetFile - relative path to the changeset file in the project
+   * @param dataSource - database to operate on
+   */
+  public void upgrade(String changesetFile, DataSource dataSource) {
+    migrateDatabase(changesetFile, dataSource, false);
+  }
+
+  private void migrateDatabase(String changesetFile, DataSource dataSource, boolean initialize) {
+    try (Connection connection = dataSource.getConnection()) {
+      Liquibase liquibase =
+          new Liquibase(
+              changesetFile, new ClassLoaderResourceAccessor(), new JdbcConnection(connection));
+      DatabaseChangeLogLock[] locks = liquibase.listLocks();
+      for (DatabaseChangeLogLock lock : locks) {
+        logger.info(
+            String.format(
+                "DatabaseChangeLogLock changeSet: %s, id: %s, lockedBy: %s, granted: %s",
+                changesetFile, lock.getId(), lock.getLockedBy(), lock.getLockGranted()));
+
+        // We can get into this state where one of the APIs is running migrations and gets shut down
+        // so that
+        // another API container can run. It will result in a lock that doesn't get released. This
+        // is similar
+        // to the problems we will have from deploying multiple containers at once that try to run
+        // migrations.
+        logger.warn("Forcing lock release");
+        liquibase.forceReleaseLocks();
```

Comment:
I don't completely follow this part. If two pods both try to run migrations, will the second one force release the lock that the first one took out? (I'm probably missing some context on how/when this gets called, so feel free to just point me somewhere I should look/read up. Thanks.)

Reply:
This is the pattern in all existing services:
https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/upgrade/Migrate.java#L338
, https://github.com/DataBiosphere/terra-workspace-manager/blob/dev/src/main/java/bio/terra/workspace/service/migrate/MigrateService.java
I don't know why we have this at beginning, but I assume this is related with the javadoc above? `"We expect MC Terra services to scale horizontally and to perform rolling upgrades of the running instances"`

Reply:
Okay, I'm not sure I follow that. But I understand you're just copying code to centralize it here, so this question should probably be directed to the groups that are using that code now.